watcher/local/adjudicatorpubsub.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"github.com/pkg/errors"
+
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/pkg/sync/atomic"
+	"perun.network/go-perun/watcher"
+)
+
+var _ watcher.AdjudicatorSub = &adjudicatorPubSub{}
+
+const adjPubSubBufferSize = 10
+
+type (
+	adjudicatorPubSub struct {
+		closed atomic.Bool
+		pipe   chan channel.AdjudicatorEvent
+	}
+
+	adjudicatorPub interface {
+		publish(channel.AdjudicatorEvent)
+		close() error
+	}
+)
+
+func newAdjudicatorEventsPubSub() *adjudicatorPubSub {
+	return &adjudicatorPubSub{
+		pipe: make(chan channel.AdjudicatorEvent, adjPubSubBufferSize),
+	}
+}
+
+func (a *adjudicatorPubSub) publish(e channel.AdjudicatorEvent) {
+	a.pipe <- e
+}
+
+// Close closes the publisher instance and all the subscriptions associated
+// with it. Any further call to Publish should immediately return error.
+func (a *adjudicatorPubSub) close() error {
+	if !a.closed.TrySet() {
+		return errors.WithStack(errAlreadyClosed)
+	}
+	close(a.pipe)
+	return nil
+}
```

Comment:
**Short answer:** It possible, I have updated the implementation to use `sync.Once` and updated `close` to return nothing. 

**Explanation:**
As mentioned in the earlier comment with @ggwpez, the reason for using `atomic.Bool` was to return an error when 
1. `close` is called more than once.
2. `Err` method is called after the subscription is closed.

But, thinking through how this error should be handled, I realized
1. Calling `close` multiple times is an acceptable behaviour and the caller only wants to ensure the pub-sub is closed after the call returns. They need not know, if they closed it or it was already closed.

    Hence, there is no need for close to return an error.
2. Calling `Err` after a subscription is closed. In case of local watcher, a subscription is closed only when
    1. Watcher crashes. In that case, client will also have crashed. So error handling is irrelevant.
    2. Client calls `StopWatching` for the channel and the `Watcher` closes the subscription. In this case, we previously returned an `errAlreadyClosed`.
    
    But, I think it is more appropriate to return `nil`, because the subscription was closed on the request of the client and no errors were encountered when closing it. Hence, updated `Err` to always return `nil`.